iceberg-rust/src/table/transaction/operation.rs

```diff
@@ -140,19 +161,26 @@ impl Operation {
 
                 let n_splits = manifest_list_writer.n_splits(n_data_files + n_delete_files);
 
-                let new_datafile_iter =
-                    delete_files
-                        .into_iter()
-                        .chain(data_files.into_iter())
-                        .map(|data_file| {
-                            ManifestEntry::builder()
-                                .with_format_version(table_metadata.format_version)
-                                .with_status(Status::Added)
-                                .with_data_file(data_file)
-                                .build()
-                                .map_err(crate::spec::error::Error::from)
-                                .map_err(Error::from)
-                        });
+                let new_datafile_iter = delete_files.into_iter().chain(data_files.into_iter()).map(
+                    |(data_file, dsn_offset)| {
+                        let mut builder = ManifestEntry::builder();
+                        builder
+                            .with_format_version(table_metadata.format_version)
+                            .with_status(Status::Added)
+                            .with_data_file(data_file);
+                        // If there is only one data sequence number in this commit, we can just use sequence number inheritance
+                        // If there are multiple data sequence numbers in this commit, we need to set the data sequence number on each manifest
+                        if multiple_data_sequence_numbers {
+                            builder.with_sequence_number(
+                                table_metadata.last_sequence_number + dsn_offset,
```

Comment:
`table_metadata.last_sequence_number` is the snapshot sequence number, which is incremented by 1 each time https://github.com/splitgraph/iceberg-rust/blob/bbfdd310b7c294d2d3f2f7ca81ba77eb82a103ce/iceberg-rust/src/table/transaction/operation.rs#L187

But here we also offset it by dsn_offset for reach DSN group, meaning DSN groups from different snapshots will end up having overlapping DSNs.

Reply:
That's one way to fix it, not sure if it's frowned upon to have discontinuous snapshot sequence numbers.

If it is, then i guess those two need to be decoupled, and some sort of look-back mechanism should be used (i.e. start of from the max DSN + 1 in the previous snapshot(s)).